cumulusci/tasks/bulkdata/dates.py

```diff
@@ -0,0 +1,115 @@
+from datetime import date, datetime
+from typing import List, Optional
+from cumulusci.core.config.OrgConfig import OrgConfig
+from cumulusci.tasks.bulkdata.step import DataOperationType
+
+
+def get_relative_date_context(mapping, org_config: OrgConfig):
+    fields = mapping.get_field_list()
+
+    date_fields = [
+        fields.index(f)
+        for f in mapping.get_fields_by_type("date", org_config)
+        if f in mapping.fields
+    ]
+    date_time_fields = [
+        fields.index(f)
+        for f in mapping.get_fields_by_type("datetime", org_config)
+        if f in mapping.fields
+    ]
+
+    return (date_fields, date_time_fields)
+
+
+def adjust_relative_dates(
+    mapping,
+    context,
+    record: List[Optional[str]],
+    operation: DataOperationType,
+):
+    """Convert specified date and time fields (in ISO format) relative to the present moment.
+    If some date is 2020-07-30, anchor_date is 2020-07-23, and today's date is 2020-09-01,
+    that date will become 2020-09-07 - the same position in the timeline relative to today."""
+
+    date_fields, date_time_fields = context
+
+    # Determine the direction in which we are converting.
+    # For extracts, we convert the date from today-anchored to mapping.anchor_date-anchored.
+    # For loads, we do the reverse.
+
+    if operation is DataOperationType.QUERY:
+        current_anchor = date.today()
+        target_anchor = mapping.anchor_date
+    else:
+        current_anchor = mapping.anchor_date
+        target_anchor = date.today()
+
+    r = record.copy()
+
+    for index in date_fields:
+        if operation is DataOperationType.QUERY:
+            index += 1  # For the Id field.
```

Comment:
This arithmetic to target the right column looks error prone. This would be a bigger effort and doesn't need to happen right now, but I wonder if we could use a different internal representation of rows so that they are still stored compactly but can be addressed by name. i.e. something like namedtuple or dataclasses

Reply:
Agreed. This caused me no end of headaches while implementing REST API support. I'd definitely like to revisit.

On an allied but much larger front, I'd like to seriously look at deprecating oid-as-pk mode. Paul's making autopks a lot more efficient, and having two modes adds a lot of surface area for wrong-expectation mistakes.